### PR TITLE
[WIP] Partially re-apply 2fbe4a8 using @__future

### PR DESCRIPTION
**Pull request by leandro-lucarella-sociomantic**
I'm not sure when @__future will be available and how is the dependency
between druntime and dmd usually handled, but I'll open this PR just to
bring the topic back to the table.

/cc @mihails-strasuns-sociomantic @mathias-lang-sociomantic

### CODE REVIEW

File: src/object.d [hunk 1]
```diff
@@ -1740,6 +1740,16 @@ class Throwable : Object
...
+
+    /**
+     * Get the message describing the error.
+     * Base behavior will be to return the `Throwable.msg` field.
+     * Override to return some other error message.
+     *
+     * Returns:
+     *  Error message
+     */
+    @__fututre const(char)[] message() const;
```
Comment (nemanja-boric-sociomantic) [inline src/object.d]: Typo `fututre`

Comment (leandro-lucarella-sociomantic) [inline src/object.d]: :man_facepalming:

**Comment by nemanja-boric-sociomantic** [comment]
It should be already available in dmd's master: https://github.com/dlang/dmd/pull/6494

**Comment by nemanja-boric-sociomantic** [comment]
Needs release notes.

**Comment by leandro-lucarella-sociomantic** [comment]
```
../dmd/generated/*/release/32/dmd -lib -ofgenerated/freebsd/release/32/libdruntime.a -Xfdruntime.json -conf= -Isrc -Iimport -w -dip1000 -m32 -fPIC  -O -release -inline  src/object.d src/core/atomic.d src/core/attribute.d src/core/bitop.d src/core/checkedint.d src/core/cpuid.d src/core/demangle.d src/core/exception.d src/core/math.d src/core/memory.d src/core/runtime.d src/core/simd.d src/core/thread.d src/core/time.d src/core/vararg.d src/core/internal/abort.d src/core/internal/convert.d src/core/internal/hash.d src/core/internal/spinlock.d src/core/internal/string.d src/core/internal/traits.d src/core/stdc/assert_.d src/core/stdc/complex.d src/core/stdc/config.d src/core/stdc/ctype.d src/core/stdc/errno.d src/core/stdc/fenv.d src/core/stdc/float_.d src/core/stdc/inttypes.d src/core/stdc/limits.d src/core/stdc/locale.d src/core/stdc/math.d src/core/stdc/signal.d src/core/stdc/stdarg.d src/core/stdc/stddef.d src/core/stdc/stdint.d src/core/stdc/stdio.d src/core/stdc/stdlib.d src/core/stdc/string.d src/core/stdc/time.d src/core/stdc/wchar_.d src/core/sync/barrier.d src/core/sync/condition.d src/core/sync/config.d src/core/sync/exception.d src/core/sync/mutex.d src/core/sync/rwmutex.d src/core/sync/semaphore.d src/core/sys/freebsd/execinfo.d src/core/sys/freebsd/sys/event.d src/core/sys/linux/stdio.d src/core/sys/linux/tipc.d src/core/sys/linux/sys/inotify.d src/core/sys/linux/sys/mman.d src/core/sys/linux/sys/signalfd.d src/core/sys/linux/sys/socket.d src/core/sys/linux/sys/sysinfo.d src/core/sys/linux/sys/time.d src/core/sys/linux/sys/xattr.d src/core/sys/linux/sys/netinet/tcp.d src/core/sys/posix/dirent.d src/core/sys/posix/signal.d src/core/sys/posix/netdb.d src/core/sys/posix/netinet/in_.d src/core/sys/posix/arpa/inet.d src/core/sys/posix/sys/ioctl.d src/core/sys/posix/sys/ipc.d src/core/sys/posix/sys/mman.d src/core/sys/posix/sys/resource.d src/core/sys/posix/sys/select.d src/core/sys/posix/sys/shm.d src/core/sys/posix/sys/socket.d src/core/sys/posix/sys/stat.d src/core/sys/posix/sys/statvfs.d src/core/sys/posix/sys/time.d src/core/sys/posix/sys/types.d src/core/sys/posix/sys/uio.d src/core/sys/posix/sys/un.d src/core/sys/posix/sys/utsname.d src/core/sys/posix/sys/wait.d src/core/sys/solaris/sys/priocntl.d src/core/sys/solaris/sys/types.d src/core/sys/solaris/sys/procset.d src/core/sys/windows/accctrl.d src/core/sys/windows/aclapi.d src/core/sys/windows/aclui.d src/core/sys/windows/basetsd.d src/core/sys/windows/basetyps.d src/core/sys/windows/cderr.d src/core/sys/windows/cguid.d src/core/sys/windows/com.d src/core/sys/windows/comcat.d src/core/sys/windows/commctrl.d src/core/sys/windows/commdlg.d src/core/sys/windows/core.d src/core/sys/windows/cpl.d src/core/sys/windows/cplext.d src/core/sys/windows/custcntl.d src/core/sys/windows/dbghelp.d src/core/sys/windows/dbghelp_types.d src/core/sys/windows/dbt.d src/core/sys/windows/dde.d src/core/sys/windows/ddeml.d src/core/sys/windows/dhcpcsdk.d src/core/sys/windows/dlgs.d src/core/sys/windows/dll.d src/core/sys/windows/docobj.d src/core/sys/windows/errorrep.d src/core/sys/windows/exdisp.d src/core/sys/windows/exdispid.d src/core/sys/windows/httpext.d src/core/sys/windows/idispids.d src/core/sys/windows/imagehlp.d src/core/sys/windows/imm.d src/core/sys/windows/intshcut.d src/core/sys/windows/ipexport.d src/core/sys/windows/iphlpapi.d src/core/sys/windows/ipifcons.d src/core/sys/windows/iprtrmib.d src/core/sys/windows/iptypes.d src/core/sys/windows/isguids.d src/core/sys/windows/lm.d src/core/sys/windows/lmaccess.d src/core/sys/windows/lmalert.d src/core/sys/windows/lmapibuf.d src/core/sys/windows/lmat.d src/core/sys/windows/lmaudit.d src/core/sys/windows/lmbrowsr.d src/core/sys/windows/lmchdev.d src/core/sys/windows/lmconfig.d src/core/sys/windows/lmcons.d src/core/sys/windows/lmerr.d src/core/sys/windows/lmerrlog.d src/core/sys/windows/lmmsg.d src/core/sys/windows/lmremutl.d src/core/sys/windows/lmrepl.d src/core/sys/windows/lmserver.d src/core/sys/windows/lmshare.d src/core/sys/windows/lmsname.d src/core/sys/windows/lmstats.d src/core/sys/windows/lmsvc.d src/core/sys/windows/lmuse.d src/core/sys/windows/lmuseflg.d src/core/sys/windows/lmwksta.d src/core/sys/windows/lzexpand.d src/core/sys/windows/mapi.d src/core/sys/windows/mciavi.d src/core/sys/windows/mcx.d src/core/sys/windows/mgmtapi.d src/core/sys/windows/mmsystem.d src/core/sys/windows/msacm.d src/core/sys/windows/mshtml.d src/core/sys/windows/mswsock.d src/core/sys/windows/nb30.d src/core/sys/windows/nddeapi.d src/core/sys/windows/nspapi.d src/core/sys/windows/ntdef.d src/core/sys/windows/ntdll.d src/core/sys/windows/ntldap.d src/core/sys/windows/ntsecapi.d src/core/sys/windows/ntsecpkg.d src/core/sys/windows/oaidl.d src/core/sys/windows/objbase.d src/core/sys/windows/objfwd.d src/core/sys/windows/objidl.d src/core/sys/windows/objsafe.d src/core/sys/windows/ocidl.d src/core/sys/windows/odbcinst.d src/core/sys/windows/ole.d src/core/sys/windows/ole2.d src/core/sys/windows/ole2ver.d src/core/sys/windows/oleacc.d src/core/sys/windows/oleauto.d src/core/sys/windows/olectl.d src/core/sys/windows/olectlid.d src/core/sys/windows/oledlg.d src/core/sys/windows/oleidl.d src/core/sys/windows/pbt.d src/core/sys/windows/powrprof.d src/core/sys/windows/prsht.d src/core/sys/windows/psapi.d src/core/sys/windows/rapi.d src/core/sys/windows/ras.d src/core/sys/windows/rasdlg.d src/core/sys/windows/raserror.d src/core/sys/windows/rassapi.d src/core/sys/windows/reason.d src/core/sys/windows/regstr.d src/core/sys/windows/richedit.d src/core/sys/windows/richole.d src/core/sys/windows/rpc.d src/core/sys/windows/rpcdce.d src/core/sys/windows/rpcdce2.d src/core/sys/windows/rpcdcep.d src/core/sys/windows/rpcndr.d src/core/sys/windows/rpcnsi.d src/core/sys/windows/rpcnsip.d src/core/sys/windows/rpcnterr.d src/core/sys/windows/schannel.d src/core/sys/windows/secext.d src/core/sys/windows/security.d src/core/sys/windows/servprov.d src/core/sys/windows/setupapi.d src/core/sys/windows/shellapi.d src/core/sys/windows/shldisp.d src/core/sys/windows/shlguid.d src/core/sys/windows/shlobj.d src/core/sys/windows/shlwapi.d src/core/sys/windows/snmp.d src/core/sys/windows/sql.d src/core/sys/windows/sqlext.d src/core/sys/windows/sqltypes.d src/core/sys/windows/sqlucode.d src/core/sys/windows/sspi.d src/core/sys/windows/stacktrace.d src/core/sys/windows/stat.d src/core/sys/windows/subauth.d src/core/sys/windows/threadaux.d src/core/sys/windows/tlhelp32.d src/core/sys/windows/tmschema.d src/core/sys/windows/unknwn.d src/core/sys/windows/uuid.d src/core/sys/windows/vfw.d src/core/sys/windows/w32api.d src/core/sys/windows/winbase.d src/core/sys/windows/winber.d src/core/sys/windows/wincon.d src/core/sys/windows/wincrypt.d src/core/sys/windows/windef.d src/core/sys/windows/windows.d src/core/sys/windows/winerror.d src/core/sys/windows/wingdi.d src/core/sys/windows/winhttp.d src/core/sys/windows/wininet.d src/core/sys/windows/winioctl.d src/core/sys/windows/winldap.d src/core/sys/windows/winnetwk.d src/core/sys/windows/winnls.d src/core/sys/windows/winnt.d src/core/sys/windows/winperf.d src/core/sys/windows/winreg.d src/core/sys/windows/winsock2.d src/core/sys/windows/winspool.d src/core/sys/windows/winsvc.d src/core/sys/windows/winuser.d src/core/sys/windows/winver.d src/core/sys/windows/wtsapi32.d src/core/sys/windows/wtypes.d src/gc/bits.d src/gc/config.d src/gc/gcinterface.d src/gc/os.d src/gc/pooltable.d src/gc/proxy.d src/gc/impl/conservative/gc.d src/gc/impl/manual/gc.d src/rt/aApply.d src/rt/aApplyR.d src/rt/aaA.d src/rt/adi.d src/rt/alloca.d src/rt/arrayassign.d src/rt/arraybyte.d src/rt/arraycast.d src/rt/arraycat.d src/rt/arraydouble.d src/rt/arrayfloat.d src/rt/arrayint.d src/rt/arrayreal.d src/rt/arrayshort.d src/rt/cast_.d src/rt/cmath2.d src/rt/config.d src/rt/cover.d src/rt/critical_.d src/rt/deh.d src/rt/deh_win32.d src/rt/deh_win64_posix.d src/rt/dmain2.d src/rt/dwarfeh.d src/rt/invariant.d src/rt/lifetime.d src/rt/llmath.d src/rt/memory.d src/rt/memset.d src/rt/minfo.d src/rt/monitor_.d src/rt/obj.d src/rt/profilegc.d src/rt/qsort.d src/rt/sections.d src/rt/sections_android.d src/rt/sections_elf_shared.d src/rt/sections_osx_x86.d src/rt/sections_osx_x86_64.d src/rt/sections_solaris.d src/rt/sections_win32.d src/rt/sections_win64.d src/rt/switch_.d src/rt/tlsgc.d src/rt/trace.d src/rt/tracegc.d src/rt/unwind.d src/rt/backtrace/dwarf.d src/rt/backtrace/elf.d src/rt/util/array.d src/rt/util/hash.d src/rt/util/random.d src/rt/util/typeinfo.d src/rt/util/utf.d src/rt/util/container/array.d src/rt/util/container/common.d src/rt/util/container/hashtab.d src/rt/util/container/treap.d src/rt/typeinfo/ti_Acdouble.d src/rt/typeinfo/ti_Acfloat.d src/rt/typeinfo/ti_Acreal.d src/rt/typeinfo/ti_Adouble.d src/rt/typeinfo/ti_Afloat.d src/rt/typeinfo/ti_Ag.d src/rt/typeinfo/ti_Aint.d src/rt/typeinfo/ti_Along.d src/rt/typeinfo/ti_Areal.d src/rt/typeinfo/ti_Ashort.d src/rt/typeinfo/ti_byte.d src/rt/typeinfo/ti_C.d src/rt/typeinfo/ti_cdouble.d src/rt/typeinfo/ti_cent.d src/rt/typeinfo/ti_cfloat.d src/rt/typeinfo/ti_char.d src/rt/typeinfo/ti_creal.d src/rt/typeinfo/ti_dchar.d src/rt/typeinfo/ti_delegate.d src/rt/typeinfo/ti_double.d src/rt/typeinfo/ti_float.d src/rt/typeinfo/ti_idouble.d src/rt/typeinfo/ti_ifloat.d src/rt/typeinfo/ti_int.d src/rt/typeinfo/ti_ireal.d src/rt/typeinfo/ti_long.d src/rt/typeinfo/ti_n.d src/rt/typeinfo/ti_ptr.d src/rt/typeinfo/ti_real.d src/rt/typeinfo/ti_short.d src/rt/typeinfo/ti_ubyte.d src/rt/typeinfo/ti_ucent.d src/rt/typeinfo/ti_uint.d src/rt/typeinfo/ti_ulong.d src/rt/typeinfo/ti_ushort.d src/rt/typeinfo/ti_void.d src/rt/typeinfo/ti_wchar.d src/etc/linux/memoryerror.d generated/freebsd/release/32/errno_c.o generated/freebsd/release/32/bss_section.o generated/freebsd/release/32/threadasm.o
gmake: *** [generated/freebsd/release/32/libdruntime.a] Segmentation fault: 11 (core dumped)
```
:thinking:

**Comment by nemanja-boric-sociomantic** [comment]
I can reproduce linux-64 failures locally, I'll take a look.

**Comment by nemanja-boric-sociomantic** [comment]
https://github.com/dlang/dmd/pull/7058

File: src/object.d [hunk 2]
```diff
@@ -1740,6 +1740,16 @@ class Throwable : Object
...
+
+    /**
+     * Get the message describing the error.
+     * Base behavior will be to return the `Throwable.msg` field.
+     * Override to return some other error message.
+     *
+     * Returns:
+     *  Error message
+     */
+    @__future const(char)[] message() const;
```
Comment (nemanja-boric-sociomantic) [inline src/object.d]: You also need default implementation here.

Comment (leandro-lucarella-sociomantic) [inline src/object.d]: OK, I guess this is one of the places where the implementation doesn't match the DIP... :unamused:

**Comment by leandro-lucarella-sociomantic** [comment]
Updated to workaround the current non-conforming DIP implementation + DMD bug.

File: src/object.d [hunk 3]
```diff
@@ -1740,6 +1740,19 @@ class Throwable : Object
...
+
+    /**
+     * Get the message describing the error.
+     * Base behavior is to return the `Throwable.msg` field.
+     * Override to return some other error message.
+     *
+     * Returns:
+     *  Error message
+     */
+    const(char)[] message() const @__future
```
Comment (Burgos) [inline src/object.d]: The way how it's implemented in DMD I think adding `@__future` to the right side is not valid. I'll double check.

Comment (Burgos) [inline src/object.d]: Yup, this doesn't work:

Compare this:

```
➜  arch tmp cat testfuture.d
class A
{
	@__future void foo() {}
}

class B: A
{
	void foo() {}
}

void main() {}
➜  arch tmp ../dmd/src/dmd testfuture.d
testfuture.d(8): Deprecation: @future base class method testfuture.A.foo is being overridden by testfuture.B.foo; rename the latter
```

to this

```
➜  arch tmp cat testfuture.d
class A
{
	void foo() @__future {}
}

class B: A
{
	void foo() {}
}

void main() {}
➜  arch tmp ../dmd/src/dmd testfuture.d
testfuture.d(8): Error: cannot implicitly override base class method testfuture.A.foo with testfuture.B.foo; add 'override' attribute
```

**Comment by Burgos** [comment]
Ok, dmd PR got merged, you can revert to the previous position of the `@__future` and the tests should pass. I would also add a compilable test with two classes that inherit `Throwable`, one defining `message` and one not.

**Comment by leandro-lucarella-sociomantic** [comment]
Updated. I added a unit test.

File: src/object.d [hunk 4]
```diff
@@ -3583,3 +3596,21 @@ unittest
...
+{
+    // Make sure basic stuff works with future Throwable.message
+    class NoMessage : Throwable
+    {
+    }
+    auto nm = new NoMessage;
+
+    class WithMessage : Throwable
+    {
+        override const(char)[] message() const
```
Comment (nemanja-boric-sociomantic) [inline src/object.d]: I suggest one more class with `const(char)[] message() const` (without `override` - it should still work modulo deprecation). If you put this into `test/` suite you can inspect the stderr output and make sure the deprecation is triggered.

Comment (nemanja-boric-sociomantic) [inline src/object.d]: You're missing constructors here:

```
src/object.d(3603): Error: class object.__unittestL3600_53.NoMessage cannot implicitly generate a default ctor when base class object.Throwable is missing a default ctor
src/object.d(3608): Error: class object.__unittestL3600_53.WithMessage cannot implicitly generate a default ctor when base class object.Throwable is missing a default ctor
```

File: src/object.d [hunk 5]
```diff
@@ -3583,3 +3596,21 @@ unittest
...
+        {
+            return "I have a custom message!";
+        }
+    }
+    auto nm = new NoMessage;
```
Comment (nemanja-boric-sociomantic) [inline src/object.d]: `auto wm = new WithMessage`

File: src/object.d (hunk 4 above)
Comment (leandro-lucarella-sociomantic) [inline src/object.d]: Damn, it's been too long since last time I did some real D programming :P

Comment (nemanja-boric-sociomantic) [inline src/object.d]: Haha :-)

**Comment by leandro-lucarella-sociomantic** [comment]
Let's see what I have done wrong next...

**Comment by nemanja-boric-sociomantic** [comment]
https://github.com/dlang/druntime/pull/1895

**Pull request closed without merging**